jax/core.py

```diff
@@ -1216,6 +1239,30 @@ def canonicalize_shape(shape):
             "smaller subfunctions.")
   raise TypeError(msg.format(shape))
 
+def sorted_named_shape_tuples(named_shape_tuples):
+  return sorted(set(named_shape_tuples),
+                key=lambda x: x if isinstance(x, str) else hash(x))
```

Comment:
Does removing the custom key function and simply using `sorted` make sense here? It will sort the string-named items fine—even ordering duplicates consistently—and similarly order all other items by their default ordering behavior (whereas their `hash`, if it exists, may not).

Also, maybe rename `sorted_unique_named_shape_items`, or simply inline the expression if there's no custom key? Up to you.

Reply:
This doesn't work for the dummy axis name objects, which don't have comparison defined (and maybe shouldn't?). I think what we want here is not so much a sorted order but a canonical order, and when the names are strings it doesn't hurt for them to be ordered alphabetically rather than by hash.

Reply:
I see. Since consistent ordering is a requirement, I tend to think we should make it explicit. We can do that in a future change if that's easier. One option is to restrict what's allowed as a name—say, only strings plus some specific type (for the generated names). Ordering by `hash` is a fine workaround for now, provided it works for the generated names in particular, but it reveals an issue with accepting anything at all as a name.

Also, if you're trying to catch string names, should it be `x if isinstance(x[0], str) ...`?